File: 0133_Clone Graph/CloneGraph_BFS.cs
public class Solution {
    public Node CloneGraph(Node node) {
        if(node == null) return null;
        var dict = new Dictionary<Node, Node>();
        dict.Add(node, new Node(node.val));
        var queue = new Queue<Node>();
        queue.Enqueue(node);
        
        while(queue.Count > 0){
            var tmpNode = queue.Dequeue();
            
            foreach(var childNode in tmpNode.neighbors){
                if(!dict.ContainsKey(childNode)){
                dict.Add(childNode, new Node(childNode.val));
                queue.Enqueue(childNode);
            }
            
                dict[tmpNode].neighbors.Add(dict[childNode]);
            }
        }
        
        return dict[node];
        
    }
}

File: 0133_Clone Graph/CloneGraph_BFS.java
class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        Map<Node, Node> map = new HashMap<>();
        Queue<Node> queue = new LinkedList<>();
        queue.offer(node);
        map.put(node, new Node(node.val));
        
        while(queue.size() > 0) {
            Node n = queue.poll();
            Node copy = map.get(n);
            for(Node next : n.neighbors) {
                if(!map.containsKey(next)) {
                    map.put(next, new Node(next.val));
                    queue.offer(next);
                }
                
                 copy.neighbors.add(map.get(next));
            }
        }
        
        return map.get(node);
    }
}